### ISSUE TITLE
API Proposal: Regex.TryMatch

### ISSUE THREAD

**Issue by anurse** [dotnet/corefx#39622]
I find myself doing this a lot with `Regex`es:

```csharp
// Some example regex. Don't judge me!!
var myRegex = new Regex(@"^https?://(www\.)?github.com/(?<owner>[^/]+)/(?<repo>[^/]+)/?.*$");

var match = myRegex.Match(inputString);
if (match.Success)
{
    var owner = match.Groups["owner"].Value;
    // ... work with the match
}
else
{
    // Do something else
}
```

I think this could be made a little more ergonomic through a new method on Regex, TryMatch. With such a method, this code could become:

```csharp
// Some example regex. Don't judge me!!
var myRegex = new Regex(@"^https?://(www\.)?github.com/(?<owner>[^/]+)/(?<repo>[^/]+)/?.*$");

if (myRegex.TryMatch(inputString, out var match))
{
    var owner = match.Groups["owner"].Value;
    // ... work with the match
}
else
{
    // Do something else
}
```

Yes, it's only a small change (one line), but it feels more _ergonomic_ to me. Before `out var` support in C#, this would've been just as clumsy as the current model, but now that we have `out var` this syntax feels like a valuable and simple addition.

I get that "ergonomics" and "aesthetics" are subjective measures though, so would welcome other input here.

## Proposed API

```csharp
public class Regex
{
	// Copied from the overloads of Match.
	public bool TryMatch(string input, out Match match);
	public bool TryMatch(string input, int startat, out Match match);
	public bool TryMatch(string input, int beginning, int length, out Match match);

	// Since the static Match methods already exist, adding a static TryMatch would be consistent. They have different args though.
	// I feel less strongly about these since I avoid the static methods on Regex.
	public static bool TryMatch(string input, string pattern, out Match match);
	public static bool TryMatch(string input, string pattern, RegexOptions options, out Match match);
	public static bool TryMatch(string input, string pattern, RegexOptions options, TimeSpan matchTimeout, out Match match);
}
```

## Implementation

It's trivial to implement this, here's an example set of extension methods I use:

```csharp
public static class RegexExtensions
{
    public static bool TryMatch(this Regex regex, string input, out Match match)
    {
        match = regex.Match(input);
        return match.Success;
    }

    public static bool TryMatch(this Regex regex, string input, int startat, out Match match)
    {
        match = regex.Match(input, startat);
        return match.Success;
    }

    public static bool TryMatch(this Regex regex, string input, int beginning, int length, out Match match)
    {
        match = regex.Match(input, beginning, length);
        return match.Success;
    }
}
```

Note that in this implementation, the value of `match` is *valid* in both branches of an `if(TryMatch(...)) {} else {}` because the match object is always returned, even in the case of an unsuccessful match. I'm certainly open to having a discussion on that. I don't feel strongly either way.

In the end, I have my extension methods so I can do just fine with those, but I think this could be useful and more discoverable if it was in the BCL.

I'd be willing to submit a PR if this was approved for consideration.

**Comment by danmosemsft** [comment]
@bartonjs is this a valid use of the Try pattern? It's not avoiding an exception.

**Comment by bartonjs** [comment]
It's definitely not the Try-Parse pattern, since the non-Try version isn't throwing in the case where this version returns false.  It structurally looks like it, and that might be a disadvantage.

Since `IsMatch` is static, perhaps these Match variant methods could/should be `HasMatch`?

```C#
if (regex.HasMatch(haystack, out Match match))
{
    ...
}
else
{
    // `match` is valid here, too, but might be Match.Empty, IIRC.
}
```

**Comment by danmosemsft** [comment]
@viktorhofer

**Comment by anurse** [comment]
I don't know that `Try` is always to avoid exceptions. For example, `ChannelReader.TryRead` returning `false` does **not** indicate that `ReadAsync` would have thrown, it just tells the caller that they need to call `WaitToReadAsync` to wait for an item to appear or for the channel to signal that it's closed.

I've always internalized `Try` patterns as being ways to indicate "failure" rather than specifically that an exception would be thrown but I'm happy to defer to the framework patterns experts here. I would agree that in 90% of current cases, the non-`Try` variant does throw an exception and that's not the case here.

Regardless, I'm not too concerned about the specific name, more the API pattern. Could we get away with just expanding `IsMatch` this way?

```csharp
public class Regex
{
	// EXISTING methods
	public bool IsMatch(string input) { throw null; }
	public bool IsMatch(string input, int startat) { throw null; }
	public static bool IsMatch(string input, string pattern) { throw null; }
	public static bool IsMatch(string input, string pattern, System.Text.RegularExpressions.RegexOptions options) { throw null; }
	public static bool IsMatch(string input, string pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout) { throw null; }

	// NEW methods
	public bool IsMatch(string input, out Match match) { throw null; }
	public bool IsMatch(string input, int startat, out Match match) { throw null; }
	public static bool IsMatch(string input, string pattern, out Match match) { throw null; }
	public static bool IsMatch(string input, string pattern, System.Text.RegularExpressions.RegexOptions options, out Match match) { throw null; }
	public static bool IsMatch(string input, string pattern, System.Text.RegularExpressions.RegexOptions options, System.TimeSpan matchTimeout, out Match match) { throw null; }
}
```

I think those overloads are all safe to add (they don't cause ambiguous matches) and all they do is provide the opportunity to get the `Match` object back if the return value is true.

Looking at the fact that all the existing `IsMatch`es pass through to `Regex.Run`, [which returns a `Match` (or `null`)](https://source.dot.net/#System.Text.RegularExpressions/System/Text/RegularExpressions/Regex.cs,442), the implementation seems trivial too.

**Comment by bartonjs** [comment]
I hadn't noticed that `IsMatch` was already both static and instance.  If the type has already gone down that road then IsMatch seems reasonable.

> I don't know that Try is always to avoid exceptions.

In this case, with it being a `public bool TrySomething(string, out T)` it really looks a lot like the Try-Parse pattern, which I recommend avoiding because the semantics aren't really the same.

**Comment by anurse** [comment]
Fair point about the pattern, as I said, not something I'm super worried about.

For clarity, I'll update the original issue content with the new proposal (and include a note about the change).

**Comment by ViktorHofer** [comment]
I'm not yet convinced that the proposed API additions would make a big change. To me it looks like this would only reduce two lines of codes in typical customer code that uses Regex. We can discuss this further as we have yet another year for 5.0 changes.
